src/exo/LoopIR_scheduling.py

```diff
@@ -2564,11 +2564,147 @@ def map_e(self, e):
+            common = { key: (lhs[key]-rhs[key]) for key in lhs if key in rhs }
+            # else, negate the rhs and cat map
+            neg_rhs = { key: -rhs[key] for key in rhs }
+            return lhs | neg_rhs | common
+        elif op == '*':
+            # rhs or lhs NEEDS to be constant
+            assert len(rhs) == 1 or len(lhs) == 1
+            # multiply the other one's value by that constant
+            if len(rhs) == 1 and self.C in rhs:
+                return { key: lhs[key]*rhs[self.C] for key in lhs }
+            else:
+                assert self.C in lhs
+                return { key: rhs[key]*lhs[self.C] for key in rhs }
+        else:
+            assert False, "bad case"
+
+    def normalize_e(self, e):
+        if isinstance(e, LoopIR.Read):
+            assert e.type.is_indexable()
+            assert len(e.idx) == 0, "Indexing inside indexing does not make any sense"
+            return { e.name : 1 }
+        elif isinstance(e, LoopIR.Const):
+            return { self.C : e.val }
+        elif isinstance(e, LoopIR.USub):
+            e_map = self.normalize_e(e.arg)
+            return { key: -e_map[key] for key in e_map }
+        elif isinstance(e, LoopIR.BinOp):
+            lhs_map = self.normalize_e(e.lhs)
+            rhs_map = self.normalize_e(e.rhs)
+            return self.concat_map(e.op, lhs_map, rhs_map)
+        else:
+            assert False, ("index_start should only be called by"+
+                           f" an indexing expression. e was {e}")
+
+    def has_div_mod(self, e):
+        if isinstance(e, LoopIR.Read):
+            return False
+        elif isinstance(e, LoopIR.Const):
+            return False
+        elif isinstance(e, LoopIR.USub):
+            return self.has_div_mod(e.arg)
+        elif isinstance(e, LoopIR.BinOp):
+            if e.op == '/' or e.op == '%':
+                return True
+            else:
+                lhs = self.has_div_mod(e.lhs)
+                rhs = self.has_div_mod(e.rhs)
+                return lhs or rhs
+        else:
+            assert False, "bad case"
+
+    # Call this when e is one indexing expression
+    # e should be an indexing expression
+    def index_start(self, e):
+        assert isinstance(e, LoopIR.expr)
+        # Div and mod need more subtle handling. Don't normalize for now.
+        if self.has_div_mod(e):
+            return e
+
+        # Make a map of symbols and coefficients
+        n_map = self.normalize_e(e)
+
+        # Write back to LoopIR.expr
+        def get_loopir(key, value):
+            vconst = LoopIR.Const(value, T.int, e.srcinfo)
+            if key == self.C:
+                return vconst
+            else:
+                readkey = LoopIR.Read(key, [], e.type, e.srcinfo)
+                return LoopIR.BinOp('*', vconst, readkey, e.type, e.srcinfo)
+        
+        delete_zero = { key: n_map[key] for key in n_map if n_map[key] != 0 }
+        new_e = LoopIR.Const(0, T.int, e.srcinfo)
```

Comment:
Will this just generate gratuitous leading zeros everywhere, or does simplification take care of that for you?

Reply:
Simplification's constant folding path takes care of eliminating the leading zeros